### ISSUE TITLE
imtcp with gtls don't expose full certificate chain to clients

### ISSUE THREAD

**Issue by cvuillemez** [rsyslog/rsyslog#2547]
### Expected behavior
```
# echo | openssl s_client -connect localhost:11514 2>/dev/null |awk '/Certificate chain/,/---/'
Certificate chain
 0 s:/C=FR/O=Foo/OU=bar/CN=bar.net/emailAddress=foo@bar.net 
   i:/C=FR/O=Foo/OU=bar/CN=intermediate Server
 1 s:/C=FR/O=Foo/OU=bar/CN=intermediate Server
   i:/C=FR/O=Foo/OU=bar/CN=Root CA
---
```

### Actual behavior
```
$ echo | openssl s_client -connect localhost:11514 2>/dev/null |awk '/Certificate chain/,/---/'
Certificate chain
 0 s:/C=FR/O=Foo/OU=bar/CN=bar.net/emailAddress=foo@bar.net 
   i:/C=FR/O=Foo/OU=bar/CN=intermediate Server
---
```

### Steps to reproduce the behavior
Put the intermediate cert in /etc/pki/rsyslog/ca.pem & /etc/pki/rsyslog/cert.pem

### Environment
- rsyslog version: 8.33.1
- platform: Ubuntu 16.04
- rsyslog.conf:
```
# provides TLS TCP syslog reception
$DefaultNetstreamDriverCAFile /etc/pki/rsyslog/ca.pem
$DefaultNetstreamDriverCertFile /etc/pki/rsyslog/cert.pem
$DefaultNetstreamDriverKeyFile /etc/pki/rsyslog/key.pem

module(load="imtcp"
       streamdriver.name="gtls"
       streamdriver.mode="1"
       streamdriver.authmode="anon"
       gnutlsprioritystring="SECURE256:SECURE128:-VERS-TLS1.0:-VERS-TLS1.1:-RSA:-CAMELLIA-128-GCM:-CAMELLIA-256-GCM:-CAMELLIA-128-CBC:-CAMELLIA-256-CBC:-AES-128-CBC:-AES-256-CBC"
#       permittedpeer="*.net"
)
[...]
input(type="imtcp" port="11514" ruleset="incominglogs")
```

*Note: rsyslog documentation is available here*:
  [pre-release](http://www.rsyslog.com/doc/master/),
  [current stable release](http://www.rsyslog.com/doc/v8-stable/)

**Comment by ajrpayne** [comment]
Rsyslog 8.33.0 from from git checkout.
Debian 9.3

Finding the same issue. Only first cert in chain is being sent to client from server. Had to add the rest of the chain to trust store as work-around

**Comment by ctramnitz** [comment]
Will this only be fixed for ossl but not for gtls? ossl is still broken...

**Comment by alorbach** [comment]
I dont see how ossl is broke by this? 
The usual setup is that client and server use the same ca cert and there is no need to transport the CA cert from client to server.

**Comment by ctramnitz** [comment]
ossl is not broken by *this*, but by #3037 
But with *this* not fixed in gtls and #3037 still open, there is **no method available to use syslog-tls**
1. without a client certificate
2. and having the client verify the server certificate (when an intermediate cert is needed)

Currently, I'm running stunnel to workaround the issues...

**Comment by alorbach** [comment]
This is correct, certless communcation needs to be allowed and implemented for both, gtls and ossl transport drivers in rsyslog. @rgerhards this is something we need to discuss.

**Comment by ctramnitz** [comment]
No, you have mistaken my point, there are two different issues, but in combination they prevent syslog-tls usage:
1. gtls does present the certificate chain, which is required when an intermediate cert needs to be presented. This prevents verification of the server cert by the client (regardless of client cert usage). This is what this issue is about, but instead of fixing it in gtls changes to ossl where made (where it works), so it is still broken in gtls.
2. #3037 is specific to client cert usage. This is unique to ossl but works fine with with gtls.

Again, the term "certless" is very misleading in this context. This is not what the problem is about or anyone asked for.

**Comment by rgerhards** [comment]
ok, let's sort this a bit. Note that my memory on the gtls driver by now is pretty weak.

As far as I understand @ctramnitz:
- this works for ossl (**which implies it works in general**)
- for gtls it does not work

So to me this sounds like a bug (or at least a missing feature) in gtls. In order to align the driver functionality, we should make this work for gtls as well.

Comments?

**Comment by ctramnitz** [comment]
Yes that is correct. The bug exists exactly as described by the initial reporter and confirmed by someone else and me. As you can see all commit to "fix" this actually went into ossl, not gtls. (This would be perfectly fine if gtls is considered deprecated and you expect everyone to switch to ossl)

**Comment by rgerhards** [comment]
actually, while I would like to deprecated GnuTLS (because we cannot get proper error messages out of it), we cannot do - some distros really prefer it. So both drivers will be supported. Anyhow, we highly recommend to use openssl, as it generates much better error messages and is far less picky when it comes to certificate formats.

**Comment by ionosphere80** [comment]
I'm running into this issue. Does it relate to implementation of the GnuTLS library or a bug in the library itself?
